Add a guide to exporting events to Splunk

ptgott opened the pull request:
Closes #13158

Since there are multiple guides to exporting audit events now, this also adds a new section of the docs for these guides.

Also fixes a tiny error in the Elastic Stack guide.

ptgott committed:
Add a guide to exporting events to Splunk

Closes #13158

Since there are multiple guides to exporting audit events now, this also
adds a new section of the docs for these guides.

Also fixes a tiny error in the Elastic Stack guide.

stevenGravy commented on docs/config.json:
```diff
@@ -484,6 +474,27 @@
           ]
+        },
+        {
+          "title": "Export Teleport Audit Events",
```
This title seems to not line up with the other titles like Admin, Operations.  Perhaps Audit Events?  The Teleport part is also not required

ptgott replied:
This was my attempt to nudge this section toward #17556, but you're right that it doesn't really fit. I'll change it.

stevenGravy commented on docs/config.json:
```diff
@@ -484,6 +474,27 @@
+          "slug": "/management/export-audit-events/",
+          "entries": [
+            {
+              "title": "Export Audit Events to Fluentd/",
```
Seems an extra `/` here.

stevenGravy commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+
+## Prerequisites
+
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
```
This works for OSS and ent right?  I don't think there's anything specific that this only works for ent.

ptgott replied:
Do plugins also work for OSS? My understanding was that plugins are a commercial feature. If not, we'll need to update all of our plugin guides.

stevenGravy replied:
Yes @ptgott , they work for OSS too.  They have some different behavior in that for Access Requests plugins they'll give CLI output instead of a link to the web access requests that are ent only.

ptgott commented:
> The event plugin instrs are different in slack and elastic then all the options to fluentd. I would expect it to be the same including helm.

I didn't test the guide on Kubernetes. My thinking was that we could produce a minimal guide for running the Event Handler on a VM first, which is still valuable, then expand it to include Kubernetes instructions later.

> Good to have a troubleshooting section for these. Things like:

> expired event handler plugins
What does it mean when a plugin has expired?

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+  name: teleport-event-handler
+spec:
+  roles: ['teleport-event-handler']
+version: v2
```
Does it matter that there's a version mismatch between these two entities?

ptgott replied:
Configuration resources are versioned separately, e.g., here are all the versions used in the `master` branch of `gravitational/teleport`:

```
$  grep -REo "type [A-Z][A-Za-z]+SpecV[0-9]+ struct" ./api/types | grep -Eo "V[0-9]+" | sort | uniq -c
   4 V1
  16 V2
  17 V3
   1 V4
   1 V5
```

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
````diff
@@ -0,0 +1,783 @@
+Next, create the role:
+
+```code
+$ tctl create teleport-event-handler-impersonator.yaml
````
Should we use `-f` consistently throughout the doc?

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
````diff
@@ -0,0 +1,783 @@
+Retrieve your local user's configuration resource:
+
+```code
+$ tctl get users/$(tsh status -f json | jq -r '.active.username') > out.yaml
````
Consider replacing `out.yaml` with a more descriptive name.

alexfornuto replied:
This assumes `jq` is installed on the workstation. A fair assumption for our users, but an assumption none the less. Maybe note the req, and link to [jq](https://stedolan.github.io/jq/)?

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+
+<ScopedBlock scope={["enterprise", "oss"]}>
+
+The format of the credentials depends on whether you have set up your network to
+give the plugin direct access to the Teleport Auth Service, or if all Teleport
+clients and services connect to the Teleport Proxy Service instead.
```
Since connecting through the proxy is the preferred/suggested configuration, I suggest restructuring this phrasing so that it's seen as the primary option. For example:

```suggestion
The format of the credentials depends on whether you have set up your Teleport
cluster such that all clients and services connect to the Teleport Proxy service,
or to give the plugin direct access to the Teleport Auth Service instead.
```

I'm sure ^ could use some tweaking though.

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+
+<Details title="Finding your $SPLUNK_HOME">
+
+To find your `$SPLUNK_HOME`, run the following command to see the location your
```
```suggestion
To find your `$SPLUNK_HOME`, run the following command to see the location of your
```

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+
+View the file at the path shown in the `Loaded:` field. Your `$SPLUNK_HOME`
+will include the filepath segments in `ExecStart` before `bin`.
+
```
Just a suggestion: We could include an example `cat /lib/systemd/system/SplunkForwarder.service` and show the relevant output here.

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+retention (days)** depend on your organization's resources and practices for log
+management.
+
+Click "Save".
```
```suggestion
Click **Save**.
```

To be consistent with styling for other web UI components.

alexfornuto commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,783 @@
+
+![Create a Token](../../../../docs/img/enterprise/plugins/splunk/new-token.png)
+
+Beside **Index** select the `teleport-audit-logs` index you created earlier.
```
Does "Beside **Index**" refer to the "Select Allowed Indexes" text in the screenshot above?

ptgott replied:
I've made this clearer

ptgott commented:
I'm squashing the "optimize images" commit on my next push to clean up the history a bit

ptgott commented on docs/pages/management/export-audit-events/splunk.mdx:
````diff
@@ -0,0 +1,783 @@
+
+  </TabItem>
+
+  <TabItem options="Docker" label="From Source">
+  You must have Docker Desktop installed on your Universal Forwarder host. 
+
+  Log in to the host and execute the following commands to build the plugin:
+
+  ```code
+  $ git clone https://github.com/gravitational/teleport-plugins.git --depth 1
+  $ cd teleport-plugins/event-handler/build.assets
+  $ make build
+  ```
+
+  You can find the compiled binary within your clone of the `teleport-plugins`
+  repo, with the file path, `event-handler/build/teleport-event-handler`. Move
+  this into your `PATH`.
+
+  </TabItem>
````
I think it's useful to include separate instructions for building from source via Docker and `go build`, but you're right that both methods result in a binary and it's misleading to lump the Docker-based build step with the `docker pull` instructions. I've suggested another way to organize these `Tabs`.

ptgott committed:
Respond to PR feedback

ptgott commented:
@stevenGravy 

- I've opened [this issue](https://github.com/gravitational/teleport/issues/17980) in response to your comment on documenting OSS Teleport's plugin support. Since this issue applies to all of our plugin guides, I think we should address it separately from this PR.

- What do you think about my responses to your initial review in [this comment](https://github.com/gravitational/teleport/pull/17405#issuecomment-1284504138)?

ptgott commented:
@stevenGravy Following up on this comment: https://github.com/gravitational/teleport/pull/17405#issuecomment-1297666563

ptgott committed:
Fix linter issues

pschisa commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,791 @@
+
+Log out of your Teleport cluster and log in again to assume the new role.
+
+### Export the access plugin identity
```
I dont see any -ttl flags on the auth commands, or any mention or reloading them later.  Wouldnt this setup break after twelve hours?  Should we encouraged longer lived certs?  Machine ID might be an option here too

ptgott replied:
I've added the `--ttl` flag to the `tctl auth sign` command snippets and mentioned Machine ID in the "Next steps" section (as well as the "Next steps" section in the Elastic and Fluentd guides). Eventually, I want to change all of our plugin instructions so they use Machine ID: https://github.com/gravitational/teleport/issues/15062

pschisa commented on docs/pages/management/export-audit-events/splunk.mdx:
````diff
@@ -0,0 +1,791 @@
+```code
+$ chmod +r teleport.ca teleport.cert teleport.key
+```
+
````
What happened to using the auth.cas, auth.crt, and auth.key method described in the connect to the auth steps previously?

ptgott replied:
Looks like I got the names of the files mixed up with the paths to the TOML fields in the Event Handler config

pschisa commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,791 @@
+</TabItem>
+</Tabs>
+
+### Start the Event Handler
```
Shouldnt we recommend running the Event Handler as a systemd service so it persists reboots?

ptgott commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,791 @@
+
+- Splunk Cloud Platform or Splunk Enterprise v9.0.1 or above.
+
+- A Linux host where you will run the Teleport Event Handler plugin and Splunk
```
I've edited the text to only say "The Teleport Event Handler" instead of using these variations

ptgott commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,791 @@
+
+On the host where you are running the Universal Forwarder and the Event Handler
+plugin, execute the following command to generate a sample configuration that we
+will edit later. Replace `mytenant.teleport.sh` with the DNS name of your
```
I didn't want to include separate addresses for Cloud/Self-Hosted deployments, so I used the Cloud one. Now that we have the `Var` component though, I'll suggest using that here instead.

ptgott commented on docs/pages/management/export-audit-events/splunk.mdx:
```diff
@@ -0,0 +1,791 @@
+In order for the Event Handler plugin to forward events from your Teleport
+cluster, it needs signed credentials from the cluster's certificate authority.
+The `teleport-event-handler` user cannot request this itself, and requires
+another user to **impersonate** this account in order to request credentials.
```
In how-to guides, I've been advising against putting links in steps, since that might delay the time it takes to reach the goal of the guide. I've made it clearer that we'll describe impersonation in this section, and added a link to the impersonation guide in the "Next steps" section.

ptgott committed:
Respond to PR feedback

- Use consistent naming
- Mention the `--ttl` flag for `tctl auth sign`
- Mention Machine ID
- Add a quick Troubleshooting section
- Fix `chmod` command
- Add systemd configs (also added this to the Elastic Stack Event
  Handler guide. I didn't do this to the Fluentd guide because the
  structure of the guide is different from the Splunk guide's).

ptgott commented:
TODO: when backporting, backport https://github.com/gravitational/teleport/pull/19514 on top of my backport branch
